ext/node/zlib/mod.rs

```diff
@@ -0,0 +1,474 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::error::bad_resource_id;
+use deno_core::error::type_error;
+use deno_core::error::AnyError;
+use deno_core::op;
+use deno_core::OpState;
+use deno_core::ZeroCopyBuf;
+use libz_sys::*;
+use std::borrow::Cow;
+use std::cell::RefCell;
+use std::future::Future;
+use std::rc::Rc;
+
+mod alloc;
+
+#[inline]
+fn check(condition: bool, msg: &str) -> Result<(), AnyError> {
+  if condition {
+    Ok(())
+  } else {
+    Err(type_error(msg.to_string()))
+  }
+}
+
+struct ZlibInner {
+  dictionary: Option<Vec<u8>>,
+  err: i32,
+  flush: i32,
+  init_done: bool,
+  level: i32,
+  mem_level: i32,
+  mode: i32,
+  strategy: i32,
+  window_bits: i32,
+  write_in_progress: bool,
+  pending_close: bool,
+  gzib_id_bytes_read: u32,
+  strm: z_stream,
+}
+
+const GZIP_HEADER_ID1: u8 = 0x1f;
+const GZIP_HEADER_ID2: u8 = 0x8b;
+
+impl ZlibInner {
+  fn start_write(
```

Comment:
I realize the zlib API is not terribly ergonomic, but bifurcating the compress/decompress paths into separate functions may simplify the code. I don't think this necessarily _has_ to happen, but it would be much easier to understand if the code paths where separated in that way.